Refine the alias_attribute warning message to be more specific

ipc103 opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

We want to be clear that alias_attribute is not meant to be used for non-attributes, and provide a clear deprecation warning if it's used for associations or otherwise non-attribute methods. This adds specific deprecation warnings for overriding an attribute method that won't forward in the future, aliasing an association (a common anti-pattern), and aliasing a non-attribute method.

In addition, we add tests for some of these common cases to ensure that we get the expected deprecation warning.

- If the user has manually defined an attribute method, we need to warn them that we won't forward to that method call in the future.
- If the user is using `alias_attribute` to alias an association (a common anti-pattern), we need to warn them of that specifically. 
- Finally, we can warn if `alias_attribute` is being used with a non-attribute target. 

### Detail

We add a check to walk up the alias chain to ensure that the target hasn't been redefined by the user. If the targets are all attribute methods and the class has the attribute, we can use `define_proxy_call`. Otherwise, we determine a deprecation warning to issue. 

### Additional information

Opened separately from https://github.com/rails/rails/pull/48913 to try to keep the changes easier to reason about, but let me know if there are any conflicts I may have missed. 

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

matthewd committed:
Refine the alias_attribute warning message to be more specific

Also, follow any chained aliases when considering whether it's safe to
use the new direct-proxy implementation.

We want to be clear that alias_attribute is not meant to be used for
non-attributes, and provide a clear deprecation warning if it's used
for associations or otherwise non-attribute methods. This adds specific
deprecation warnings for overriding an attribute method that won't
forward in the future, aliasing an association (a common anti-pattern),
and aliasing a non-attribute method.

In addition, we add tests for some of these common cases to ensure that
we get the expected deprecation warning.

Co-authored-by: Ian Candy <ipc103@github.com>

rafaelfranca commented:
That is a lot of complexity just to warn users about things that are doing wrong. Can you please clean this code? It is full of comments that probably should be there but are in reality clear indication that all this logic should not be in the same method.

ipc103 committed:
Refactor steps for identifying deprecation warnings

First pass at making this logic easier to read/understand, and hopefully
also make it easier to remove when we're ready to get rid of the
deprecation warnings.

ipc103 commented:
@rafaelfranca Thanks for looking - I took a first pass at cleaning this up and will give it another tidying before re-requesting review. I definitely agree about the added logic/complexity, but hope that it will make it easier for folks to resolve the issues by providing clearer deprecation warnings. Also hopeful that all of this logic is only temporary until the old behavior is not supported.

ipc103 committed:
Extract module

ipc103 committed:
Refactor: break up walking chain with finding aliases

We can find the alias target and then issue the deprecation warning
separately to make the logic a little bit easier to reason about.

ipc103 committed:
Separate checking for deprecations from raising warnings

ipc103 committed:
Add additional warning

When you make a typo in the attribute name, we also should warn
that you probably made a mistake. This updates the logic so that we
make sure the method and attribute are present that you're trying to alias.

Also, we check that the pattern is the method_name exactly so that we only
generate that warning once. We don't need to generate these for every single
generated method, only the initial one.

ipc103 commented:
I did a bit more refactoring, but the logic was still pretty complex. I'm going to close this out in favor of #48972, which clarifies the deprecation warning in a simpler way.
